storage/Devices/MdImpl.cc

```diff
@@ -228,12 +230,27 @@ namespace storage
 	    if (!proc_mdstat.has_entry(short_name))
 		continue;
 
-	    MdadmDetail mdadm_detail = prober.get_system_info().getMdadmDetail(name);
+	    const MdadmDetail& mdadm_detail = prober.get_system_info().getMdadmDetail(name);
 	    if (!mdadm_detail.devname.empty())
 		name = DEVMDDIR "/" + mdadm_detail.devname;
 
-	    Md* md = Md::create(prober.get_probed(), name);
-	    md->get_impl().probe_pass_1a(prober);
+	    const ProcMdstat::Entry& entry = prober.get_system_info().getProcMdstat().get_entry(short_name);
+
+	    if (entry.is_container)
+	    {
+		MdContainer* md_container = MdContainer::create(prober.get_probed(), name);
+		md_container->get_impl().probe_pass_1a(prober);
+	    }
+	    else if (entry.has_container)
+	    {
+		MdMember* md_member = MdMember::create(prober.get_probed(), name);
+		md_member->get_impl().probe_pass_1a(prober);
+	    }
+	    else
+	    {
+		Md* md = Md::create(prober.get_probed(), name);
+		md->get_impl().probe_pass_1a(prober);
+	    }
```

Comment:
Base class being aware about derived ones is a bit weird. Would it be possible different probe methods for each one?